app/controllers/api/v1/resources_controller.rb

```diff
@@ -4,51 +4,91 @@ class Api::V1::ResourcesController < ApplicationController
 
   before_filter :authenticate_api_key!
   before_filter :require_allowed_model
-
-  # FIXME find a way to document api calls
+  before_filter :require_strictly_correct_params, only: :create
 
   def index
-    filter_params = params[model_name.singularize].presence
+    filter_params = params[model_name].presence
 
     resources = filter_params ? model_klass.filter(filter_params) : model_klass.all
     resources = paginate_collection(resources)
     resources = decorate_collection(resources)
 
-    render json: { model_name => resources, :meta => resources.meta }
+    render json: { model_name.pluralize => resources, :meta => resources.meta }
   end
 
   def show
     resource = model_klass.find(params[:id])
-    render json: decorate(resource)
+    render json: { model_name => decorate(resource) }
+  end
+
+  def create
+    resource = model_klass.new(create_params.merge(:user_id => api_key.user_id))
+
+    if resource.save
+      render json: { model_name => decorate(resource) }, status: :created
+    else
+      errors = []
+
+      resource.errors.messages.each do |attr, messages|
+        messages.each do |msg|
+          errors << { attribute: attr, message: msg.capitalize }
+        end
+      end
+
+      render json: { errors: errors }, status: 422
+    end
   end
 
   private
 
   def authenticate_api_key!
-    token = ApiKey.normalize_token(params[:api_key])
-    unless token.present? && ApiKey.exists?(token: params[:api_key])
+    unless api_key.present?
       render text: "Not Found", status: 404
     end
   end
 
+  def api_key
+    return @api_key if defined?(@api_key)
+
+    token = params[:api_key] || request.headers["X-BIP-Api-Key"]
+    token = ApiKey.normalize_token(token)
+
+    @api_key = token && ApiKey.find_by(token: token)
+  end
+
   def require_allowed_model
-    unless allowed_models.include?(model_name)
+    if request.request_method_symbol == :get && !Api.readable_model?(model_name)
+      raise ActionController::RoutingError.new('Not Found')
```

Comment:
Why sometimes we throw a RoutingError (that, I assume, results in a 404) and sometime (the no key case handling above) we simply render text with status = 404?